File: common/src/main/java/juuxel/adorn/util/Services.java
package juuxel.adorn.util;

import java.util.IdentityHashMap;
import java.util.Map;
import java.util.ServiceLoader;

public final class Services {
    private static final Map<Class<?>, Object> CACHE = new IdentityHashMap<>();

    @SuppressWarnings("unchecked")
    public static <T> T load(Class<T> c) {
        return (T) CACHE.computeIfAbsent(c, Services::loadService);
    }

    private static <T> T loadService(Class<T> c) {
        return ServiceLoader.load(c)
            .findFirst()
            .orElseThrow(() -> new RuntimeException("Could not find Adorn platform service " + c.getSimpleName()));
    }
}

File: common/src/main/kotlin/juuxel/adorn/util/Services.kt
package juuxel.adorn.util

import java.util.Optional
import java.util.ServiceLoader

inline fun <reified T> loadService(): T =
    ServiceLoader.load(T::class.java).findFirst().unwrapService(T::class.java)

fun <T> Optional<T>.unwrapService(type: Class<T>): T =
    orElseThrow { RuntimeException("Could not find Adorn platform service ${type.simpleName}") }
